packages/http-server-javascript/src/common/serialization/json.ts

```diff
@@ -0,0 +1,423 @@
+  switch (type.kind) {
+    case "Model": {
+      if (isArrayModelType(ctx.program, type)) {
+        const argumentType = type.indexer.value;
+
+        if (requiresJsonSerialization(ctx, argumentType)) {
+          return `${expr}.map((item) => ${transposeExpressionToJson(ctx, argumentType, "item", module)})`;
+        } else {
+          return expr;
+        }
+      } else if (isRecordModelType(ctx.program, type)) {
+        const argumentType = type.indexer.value;
+
+        if (requiresJsonSerialization(ctx, argumentType)) {
+          return `Object.fromEntries(Object.entries(${expr}).map(([key, value]) => [String(key), ${transposeExpressionToJson(
+            ctx,
+            argumentType,
+            "value",
+            module
+          )}]))`;
+        } else {
+          return expr;
+        }
+      } else if (!requiresJsonSerialization(ctx, type)) {
+        return expr;
+      } else {
+        requireSerialization(ctx, type, "application/json");
+        const typeReference = emitTypeReference(ctx, type, NoTarget, module);
+
+        return `${typeReference}.toJsonObject(${expr})`;
+      }
+    }
+    case "Scalar":
+      return expr;
+    case "Union":
+      if (!requiresJsonSerialization(ctx, type)) {
+        return expr;
+      } else {
+        requireSerialization(ctx, type, "application/json");
+        const typeReference = emitTypeReference(ctx, type, NoTarget, module, {
+          altName: "WeirdUnion",
+          requireDeclaration: true,
+        });
+
+        return `${typeReference}.toJsonObject(${expr})`;
+      }
+    case "ModelProperty":
+      // TODO/witemple - pretty sure this isn't right. If the model property has additional encode/decode parameters we
+      // should have some way to prioritize them lower than the top level property that this originated from.
+      return transposeExpressionToJson(ctx, type.type, expr, module);
+    case "Intrinsic":
+      switch (type.name) {
+        case "ErrorType":
+          throw new Error("UNREACHABLE: ErrorType in JSON deserialization");
+        case "void":
+          return "undefined";
+        case "null":
+          return "null";
+        case "never":
+        case "unknown":
+          return expr;
+        default:
+          throw new Error(
+            `Unreachable: intrinsic type ${(type satisfies never as IntrinsicType).name}`
+          );
+      }
+    case "String":
+    case "Number":
+    case "Boolean":
+      return literalToExpr(type);
+    case "Interface":
+    case "Enum":
+    case "EnumMember":
+    case "TemplateParameter":
+    case "Namespace":
+    case "Operation":
+    case "StringTemplate":
+    case "StringTemplateSpan":
+    case "Tuple":
+    case "UnionVariant":
+    case "Function":
+    case "Decorator":
+    case "FunctionParameter":
+    case "Object":
+    case "Projection":
+    case "ScalarConstructor":
+    default:
+      throw new UnimplementedError(`transformJsonExprForType: ${type.kind}`);
```

Comment:
here too

Reply:
What are you suggesting for this particular instance? I am anyways planning to rewrite almost all of this code to be more HTTP-specific. Tracking issue: https://github.com/microsoft/typespec/issues/3780